packages/react-components/react-table/src/hooks/useColumnResizeState.ts

```diff
@@ -0,0 +1,85 @@
+import { useIsomorphicLayoutEffect } from '@fluentui/react-utilities';
+import * as React from 'react';
+import {
+  TableColumnDefinition,
+  TableColumnId,
+  ColumnResizeState,
+  ColumnWidthState,
+  UseColumnSizingParams,
+} from './types';
+import {
+  columnDefinitionsToState,
+  adjustColumnWidthsToFitContainer,
+  getColumnById,
+  setColumnProperty,
+  getTotalWidth,
+  getLastColumn,
+  getColumnByIndex,
+  getColumnWidth,
+  getLength,
+} from '../utils/columnResizeUtils';
+
+export function useColumnResizeState<T>(
+  columns: TableColumnDefinition<T>[],
+  containerWidth: number,
+  params: UseColumnSizingParams = {},
+): ColumnResizeState {
+  const { onColumnResize, columnSizingOptions } = params;
+
+  const [state, setState] = React.useState<ColumnWidthState[]>(
+    columnDefinitionsToState(columns, undefined, columnSizingOptions),
+  );
+
+  // Use layout effect here to make sure that updated columns receive proper styles immediately
+  useIsomorphicLayoutEffect(() => {
+    const intermediateState = columnDefinitionsToState(columns, state, columnSizingOptions);
+    setState(adjustColumnWidthsToFitContainer(intermediateState, containerWidth));
+  }, [columns, containerWidth, state, columnSizingOptions]);
```

Comment:
```suggestion
  }, [columns, containerWidth, columnSizingOptions]);
```

From what I understand this effect should only be called as a result of column update, so it might be ok to break the rules of hooks here. `setColumnWidth` and `resetColumnWidth` already call `setState` when invoked.

Reply:
Actually, this effect is a bit odd, and really feels that we should get back to the reducer idea we discussed previously. Let's talk more about this offline